core/src/main/java/hudson/Util.java

```diff
@@ -1688,6 +1689,39 @@ public static void closeAndLogFailures(@CheckForNull Closeable toClose, @Nonnull
         }
     }
 
+    @Restricted(NoExternalUse.class)
+    public static int permissionsToMode(Set<PosixFilePermission> permissions) {
+        PosixFilePermission[] allPermissions = PosixFilePermission.values();
+        int result = 0;
+        for (int i = 0; i < allPermissions.length; i++) {
+            result <<= 1;
+            result |= permissions.contains(allPermissions[i]) ? 1 : 0;
+        }
+        return result;
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static Set<PosixFilePermission> modeToPermissions(int mode) throws IOException {
+         // Anything larger is a file type, not a permission.
+        int PERMISSIONS_MASK = 07777;
+        // setigd/setuid/sticky are not supported.
+        //                         rwxrwxrwx
+        int MAX_SUPPORTED_MODE = 0b111111111;
+        mode = mode & PERMISSIONS_MASK;
+        if ((mode & MAX_SUPPORTED_MODE) != mode) {
+            throw new IOException("Invalid mode: " + mode);
+        }
+        PosixFilePermission[] allPermissions = PosixFilePermission.values();
+        Set<PosixFilePermission> result = EnumSet.noneOf(PosixFilePermission.class);
```

Comment:
Is it covered by tests? https://docs.oracle.com/javase/7/docs/api/java/util/EnumSet.html#noneOf(java.lang.Class) is not explicit, but I am aware that some modifications can return a non-modifyable set (performance tweaks).

Reply:
There is a new test for the method (`UtilTest#testPermissionsToMode`) which should cover modification, but I will double check the coverage report. I can update to use a HashSet if you prefer.

Reply:
It is a restricted method so I am not concerned.